crates/bindings-csharp/BSATN.Codegen/Type.cs

```diff
@@ -7,327 +7,287 @@ namespace SpacetimeDB.Codegen;
+        : this(name, SymbolToName(type), GetTypeInfo(type)) { }
 
-    public MemberDeclaration(string name, ITypeSymbol typeSymbol)
-    {
-        Name = name;
-        Type = SymbolToName(typeSymbol);
-        TypeInfo = GetTypeInfo(typeSymbol);
-    }
-}
+    public MemberDeclaration(IFieldSymbol field)
+        : this(field.Name, field.Type) { }
 
-abstract record TypeKind
-{
-    public record Product : TypeKind
+    public static string GenerateBsatnFields(
+        Accessibility visibility,
+        IEnumerable<MemberDeclaration> members
+    )
     {
-        public sealed override string ToString() => "Product";
+        var visStr = SyntaxFacts.GetText(visibility);
+        return string.Join(
+            "\n",
+            members.Select(m => $"{visStr} static readonly {m.TypeInfo} {m.Name} = new();")
+        );
     }
 
-    public record Sum : TypeKind
-    {
-        public sealed override string ToString() => "Sum";
-    }
+    public static string GenerateDefs(IEnumerable<MemberDeclaration> members) =>
+        $$"""
+        new SpacetimeDB.BSATN.AggregateElement[] {
+            {{string.Join(
+                ",\n",
+                members.Select(m => $"new(nameof({m.Name}), {m.Name}.GetAlgebraicType(registrar))")
+            )}}
+        }
+        """;
+}
 
-    public record Table(bool IsScheduled) : Product;
+public enum TypeKind
+{
+    Product,
+    Sum,
 }
 
-record TypeDeclaration
+public abstract record BaseTypeDeclaration<M>
+    where M : MemberDeclaration, IEquatable<M>
 {
     public readonly Scope Scope;
     public readonly string ShortName;
     public readonly string FullName;
     public readonly TypeKind Kind;
-    public readonly EquatableArray<MemberDeclaration> Members;
+    public EquatableArray<M> Members { get; init; }
 
-    public TypeDeclaration(
-        TypeDeclarationSyntax typeSyntax,
-        INamedTypeSymbol type,
-        TypeKind kind,
-        IEnumerable<IFieldSymbol> members
-    )
+    protected abstract M ConvertMember(IFieldSymbol field);
+
+    public BaseTypeDeclaration(GeneratorAttributeSyntaxContext context)
     {
+        var typeSyntax = (TypeDeclarationSyntax)context.TargetNode;
+        var type = (INamedTypeSymbol)context.TargetSymbol;
+
+        // Note: we could use naively use `type.GetMembers()` to get all fields of the type,
+        // but some users add their own fields in extra partial declarations like this:
+        //
+        // ```csharp
+        // [SpacetimeDB.Type]
+        // partial class MyType
+        // {
+        //     public int TableField;
+        // }
+        //
+        // partial class MyType
+        // {
+        //     public int ExtraField;
+        // }
+        // ```
+        //
+        // In this scenario, only fields declared inside the declaration with the `[SpacetimeDB.Type]` attribute
+        // should be considered as BSATN fields, and others are expected to be ignored.
```

Comment:
Ah, that's brutal.

Reply:
We could either work without the need for partials, or add attributes to hide fields from codegen.